lib/Sema/CSBindings.cpp

```diff
@@ -204,14 +204,26 @@ ConstraintSystem::getPotentialBindings(TypeVariableType *typeVar) {
 
     case ConstraintKind::BridgingConversion:
     case ConstraintKind::CheckedCast:
-    case ConstraintKind::DynamicTypeOf:
     case ConstraintKind::EscapableFunctionOf:
     case ConstraintKind::OpenedExistentialOf:
     case ConstraintKind::KeyPath:
     case ConstraintKind::KeyPathApplication:
       // Constraints from which we can't do anything.
       continue;
 
+    case ConstraintKind::DynamicTypeOf: {
+      // Direct binding of the left-hand side could result
+      // in `DynamicTypeOf` failure if right-hand side is
+      // bound (because 'Bind' requires equal types to
+      // succeed), or left is bound to Any which is not an
+      // [existential] metatype.
+      if (constraint->getFirstType()->isEqual(typeVar))
+        return result;
```

Comment:
First one works just fine with the changes in the patch, the second one doesn't because `type(of:)` returns `.Type.Type` so the declaration of `foo` has to change to `foo<X>(_: X.Type.Type, _: X)` then it correctly picks `C.T`. 

The reason why it works is because in both cases there is a `arg conv` constraint present which disambiguates the type...